### ISSUE TITLE
Pick<> not working in React setState with updater function 

### ISSUE THREAD

**Issue by Suzii** [microsoft/TypeScript#16923]
**TypeScript Version:**  2.4.0 / nightly (2.5.0-dev.201xxxxx)
The React [`setState(...)` function](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L235) fails to detect incompatible value being returned from an updater function. `Pick<>` does not seem to work well in this case. I thought this would be fixed in v2.4 but it is not.

The issue migh be related to this one https://github.com/Microsoft/TypeScript/issues/13359 but it does not seem to be limited to conditional updates. 

**Code**
```ts
class MyComponent extends React.PureComponent<{}, { value: string }> {
  _onChange = (event: ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;

    this.setState({ valueWithTypo: value }); // correctly reports an error
    this.setState(() => ({ valueWithTypo: value })); // does not detect the typo
  };

  render() {
    // some implementation
  }
}
```

**Expected behavior:**
Compiler detects wrong return type of the updater function passed to `this.setState(...)`. (Overload which accepts object (not the updater function) reports an error correctly.)

**Actual behavior:**
Compiler fails to detect wrong return type of the updater function passed to `this.setState(...)`. No error reported even though argument `{ valueWithTypo: value }` is not compatible with state type `{ value: string }`.

**Comment by sylvanaar** [comment]
I have reduced your problem down to the typing that is causing the problem.

```Typescript
let f:Pick< { value: string }, never> = () => { return { a:1 }}
`

When you apply the parameter you pass in the second function, Pick<> evaluates to never, and you can assign the function to never, so it chooses the first overload.

**Comment by dakom** [comment]
On a related note, it seems the "broken" version here was the best fix for setting via a dynamic index.

Consider this snippet, modelled after the "Multiple Inputs" section at https://facebook.github.io/react/docs/forms.html but with TS where state satisfies a particular interface: 

```
onFormUpdated(evt:React.ChangeEvent<HTMLInputElement>) {
    let name = evt.currentTarget.getAttribute("name");
    let value = evt.currentTarget.value;
    this.setState(() => ({ [name]: value}));
}
```

Will this break if the issue here is fixed? Is there a cleaner way to write it?

**Comment by danthegoodman** [comment]
There's been some discussion (but no resolution) of this over in https://github.com/DefinitelyTyped/DefinitelyTyped/issues/18365
